prometheus_client/parser.py

```diff
@@ -20,128 +20,96 @@ def text_string_to_metric_families(text):
-                result.append('\\')
-            elif char == 'n':
-                result.append('\n')
-            else:
-                result.append('\\' + char)
-            slash = False
-        else:
-            if char == '\\':
-                slash = True
-            else:
-                result.append(char)
+def _replace_escaping(s):
+    return s.replace("\\n", "\n").replace('\\\\', '\\').replace('\\"', '"')
 
-    if slash:
-        result.append('\\')
 
-    return ''.join(result)
+def _parse_labels(labels_string):
+    labels = {}
+    # return if we don't have valid labels
+    if "=" not in labels_string:
+        return labels
+
+    escaping = False
+    if "\\" in labels_string:
+        escaping = True
+
+    # copy to keep original labels
+    sub_labels = labels_string
+    try:
+        # process one label at a time
+        while sub_labels:
+            # the label name is before the equal
+            value_start = sub_labels.index("=")
+            label_name = sub_labels[:value_start]
+            sub_labels = sub_labels[value_start + 1:].lstrip()
+            # find the first quote after the equal
+            quote_start = sub_labels.index('"') + 1
+            value_substr = sub_labels[quote_start:]
+
+            # find the last unescaped quote
+            i = 0
+            while i < len(value_substr):
+                i = value_substr.index('"', i)
+                if value_substr[i - 1] != "\\":
+                    break
+                i += 1
+
+            # the label value is inbetween the first and last quote
+            quote_end = i + 1
+            label_value = sub_labels[quote_start:quote_end]
+            # replace escaping if needed
+            if escaping:
+                label_value = _replace_escaping(label_value)
+            labels[label_name.strip()] = label_value.strip()
+
+            # remove the processed label from the sub-slice for next iteration
+            sub_labels = sub_labels[quote_end + 1:]
+            next_comma = sub_labels.find(",") + 1
+            sub_labels = sub_labels[next_comma:].lstrip()
+
+        return labels
+
+    except ValueError:
+        raise ValueError("Invalid labels: %s" % labels_string)
+
+
+# if we have multiple values only consider the first
+def _parse_value(s):
+    s = s.lstrip()
+    separator = " "
+    if separator not in s:
+        separator = "\t"
+    i = s.find(separator)
+    if i == -1:
+        return s
+    return s[:i]
 
 
 def _parse_sample(text):
-    name = []
-    labelname = []
-    labelvalue = []
-    value = []
-    labels = {}
-
-    state = 'name'
+    # detect the labels in the text
+    try:
+        label_start, label_end = text.index("{"), text.rindex("}")
+        # the name is before the labels
+        name = text[:label_start].strip()
+        # we ignore the starting curly brace
+        label = text[label_start + 1:label_end]
+        # the value is after the label end (ignoring curly brace and space)
```

Comment:
There can be a trailing comma after the last "

Reply:
should be already covered, `test_commas` is validating this